### ISSUE TITLE
logind: optionally, allow existance of specially marked sessions to act as inhibitors

### ISSUE THREAD

**Issue by wolfgangkarall** [systemd/systemd#20654]
**Is your feature request related to a problem? Please describe.**
When closing my notebook's lid while e.g. my SSH-based backup is running (i.e. having an active SSH session) the notebook does not go to sleep. This can easily go unnoticed ('constant glow LED' vs. 'slowly pulsating LED' on the Thinkpad lid) and hence to the notebook being powered on over long periods (in my current example: last night, ca. 14 hours) with the lid closed, possibly leading to thermal problems (at wake-up today in the morning the temperature was 75degs Celsius, which I never see under normal circumstances unless I'm compiling some software). 

**Describe the solution you'd like**
Specifically, some way to tell the inhibitor logic to ignore SSH logins when considering if the system can be put to sleep, rebooted, whatever. 

In general, the inhibitor logic seems to be an 'all or nothing' decision, as both the help text issued from `systemctl suspend`

> Alternatively, ignore inhibitors and users with 'systemctl suspend -i'.

or the documentation in logind.conf(5) 

> LidSwitchIgnoreInhibited= defaults to "yes". This means that when systemd-logind is handling events by itself (no low level inhibitor locks are taken by another application), the lid switch does not respect suspend blockers by default, but the power and sleep keys do.

suggest there's no way to selectively configure what is OK to ignore and what is not.

Please do not suggest not using pam_systemd for sshd, which I guess would resolve this issue but lead to many others, like hanging SSH sessions when issuing a reboot via SSH.

**Describe alternatives you've considered**
I considered putting a file in /lib/systemd/system-sleep/ to somehow work around the issue, but from a quick test those seem to be run after the inhibitor checks. Besides, I wouldn't really want to kill active SSH sessions to get past the inhibitor check anyway, because SSH can easily survive a 'close lid -> open lid' cycle (at times I actually realize there is an SSH session running that is important, and the SSH connection recovers even longer periods of that kind of interruption).

**The systemd version you checked that didn't have the feature you are asking for**
systemd 245, nothing related to this issue in the more recent NEWS entries.

**Comment by poettering** [comment]
i don't follow. "ssh based backups"? what does that mean? an ssh session into the local host? or out of the local host?

does that "ssh based backup" even take an inhibitor lock? if so, why? 

do you have LidSwitchIgnoreInhibited= enabled or not?

**Comment by wolfgangkarall** [comment]
Yes, "SSH-based backup is running (i.e. having an active SSH session)" means "an ssh session into the local host" in this case.

And I did not change LidSwitchIgnoreInhibited in /etc/systemd/logind.conf, nor any other property. There's no /etc/systemd/logind.conf.d/*.conf either, so I guess it's still the default 'yes'. If there's a way to check the actually set properties for logind, like with systemctl show for systemd itself, let me know.

Using gdbus I see:

```
# gdbus introspect --system --dest org.freedesktop.login1 --object-path /org/freedesktop/login1 | grep -e Lid -e Inhibit
      ListInhibitors(out a(ssssuu) arg_0);
      Inhibit(in  s arg_0,
      readonly s BlockInhibited = 'handle-power-key:handle-suspend-key:handle-hibernate-key';
      readonly s DelayInhibited = 'shutdown:sleep';
      readonly t InhibitDelayMaxUSec = 30000000;
      readonly s HandleLidSwitch = 'suspend';
      readonly s HandleLidSwitchExternalPower = '';
      readonly s HandleLidSwitchDocked = 'ignore';
      readonly b LidClosed = false;
      readonly t InhibitorsMax = 8192;
      readonly t NCurrentInhibitors = 4;
```

Not sure how I can check this "no low level inhibitor locks are taken by another application" restriction, does that mean NCurrentInhibitors would need to be 0?

As for those inhibitors:

```
# gdbus call --system --dest org.freedesktop.login1 --object-path /org/freedesktop/login1 --method org.freedesktop.login1.Manager.ListInhibitors | sed 's/),/),\n/g'
([('sleep', 'NetworkManager', 'NetworkManager needs to turn off networks', 'delay', uint32 0, uint32 866),
 ('sleep', 'UPower', 'Pause device polling', 'delay', 0, 5449),
 ('handle-power-key:handle-suspend-key:handle-hibernate-key', 'xfce4-power-manager', 'xfce4-power-manager handles these events', 'block', 1000, 5857),
 ('shutdown', 'Unattended Upgrades Shutdown', 'Stop ongoing upgrades or perform upgrades before shutdown', 'delay', 0, 1532)],)
```

**Comment by wolfgangkarall** [comment]
It would be great if the label:"needs-reporter-feedback ❓" would be removed, maybe then this issue would get some visibility again.

**Comment by sftim** [comment]
I think the thing that needs to change is either:
- the SSH daemon (to add an inhibitor when active)
- the session that the SSH daemon has started

systemd itself doesn't set inhibition

**Comment by wolfgangkarall** [comment]
Oh, the active ssh sessions _are_ inhibiting the system to go to sleep already, hence I'm told to use `systemctl suspend -i`.

It's more a matter of allowing the user to override this so that even with SSH sessions being active, the notebook can be put to sleep reliably by just closing the lid as usual.

**Comment by poettering** [comment]
I think it would be OK if we'd optionally allow a session to act as an inhibitor. This should be something one could configure in the PAM stack, either via a param to the PAM module, or via some property/env var set on the PAM session object.

But this requires a patch to be provided.

**Comment by wolfgangkarall** [comment]
Could some please tell me how this went from 'I want to make SSH sessions **not inhibit** the suspend/sleep of notebooks' to 'optionally, allow existance of speciall marked sessions **to act as inhibitors**'?

github's **bold** is way too easy to miss IMO, so in short: `not inhibit` -> `to act as inhibitors`, why? Do I misunderstand some nomenclature here or do you all miss the point of my issue?

**Comment by poettering** [comment]
hmm, what? ssh logins are already inhibiting suspend? hmm, i must have forgotten about that...

**Comment by poettering** [comment]
i didn't forget about that, because such a logic simply doesn't exist. 

do you have your laptop docked when you see this? i.e. a docking station connected or an external screen or so connect? if so, logind will detect your system as docked, and then apply "HandleLidSwitchDocked" instead of "HandleLidSwitch" and that's set to ignore for you...

what does `busctl get-property org.freedesktop.login1 /org/freedesktop/login1 org.freedesktop.login1.Manager Docked` report when this happens?

**Comment by wolfgangkarall** [comment]
Ok, if ssh logins do not inhibit sleep, why do I get this then?

```
12:15:54 [:~] $ systemctl suspend
User ... is logged in on sshd.
User ... is logged in on sshd.
Please retry operation after closing inhibitors and logging out other users.
Alternatively, ignore inhibitors and users with 'systemctl suspend -i'.
```

That busctl shows:

```
12:16:31 [:~] $ busctl get-property org.freedesktop.login1 /org/freedesktop/login1 org.freedesktop.login1.Manager Docked
b false
```

**Comment by tloimer** [comment]
> hmm, what? ssh logins are already inhibiting suspend? hmm, i must have forgotten about that...

Exactly, j'accuse, `systemctl --check-inhibitors=yes poweroff` checks (i) whether any inhibitors are taken and (ii) whether any user is logged in via sshd. Check (ii) is not documented and should not be mixed with intentionally taken inhibitors, I believe. Example, when trying to do the daily poweroff of our backup node:
```
User somebody is logged in on sshd.
Please retry operation after closing inhibitors and logging out other users.
Alternatively, ignore inhibitors and users with 'systemctl poweroff -i'.
```
I would ask for either removing the check for open ssh sessions (which would break current behaviour), or alternatively add one more option (--check-users, --check-user-sessions ?) that allows to switch off checking for logged in users.

**Comment by wolfgangkarall** [comment]
Ok, one final try, to make my initial problem understandable and reproducible, to get past this obvious confusion about 'does or does it not inhibit', and maybe (just maybe) have this hijacked issue title fixed again to something that actually corresponds to what my issue is about, as simple step-by-step instructions:

- install the most recent Fedora Server Beta (so we get a recent systemd version) into 2 VMs, I used
```
SHA256 (Fedora-Server-dvd-x86_64-40_Beta-1.10.iso) = 89ed544a677aa73ab81871b428da2ecad461edfe2ea44818416308eacb1c2712
```
from https://fedoraproject.org/server/download?beta

I'll call the initially created user `fedora` in the steps below.
- make sure you have a 2nd regular user on both of the 2 VMs, let's call it `no-admin` (because it's not in group 'wheel')
- on VM 1:
  - log in locally on the console as `no-admin`
  - ssh to VM 2
- on VM 2:
  - log in locally on the console as `fedora`
  - try to poweroff (or reboot, or suspend) using systemctl, e.g. `systemctl poweroff`

The result will be:

![image](https://github.com/systemd/systemd/assets/5254079/5abfd0b2-d9ae-40fe-9862-e614d9fcc94d)

If OTOH the same is done using the _same_ user, i.e. user `fedora` is logged in both via ssh and the local console, the system will happily kick out the ssh session and the system will be powered off (it doesn't matter if the user is in group 'wheel' or not). The same is true if user `root` is running that `systemctl poweroff` command, the host is shut down without caring about active ssh sessions (thank god for small mercies).

In my initially described scenario, the backup user connected via ssh is

a) different from the user logged in locally on the notebook, and
b) _not_ `root`,

so the same happens there, the ssh session of the backup user is inhibiting suspend when the locally logged in user closes the notebook lid.

I would very much like to configure this (maybe it is possible by now, almost 3 years after the initial report, feel free to point me in the direction of the relevant documentation), so that the local admin can say 'I do not want to ssh sessions to inhibit reboot/poweroff/suspend' if they think it makes sense for their usage scenario.

**Comment by hdante** [comment]
Hello, Wolfgang, yes, I had already understood your situation, thanks for better isolating the problem. As I had written, systemd is doing an additional check, that does not follow the inhibit protocol, for open login sessions. For the sake of agreeing on terminology, that's not an ssh session, it's a logind session. Code follows, explicitly:

```c
        /* Check for current sessions */
        sd_get_sessions(&sessions);
        STRV_FOREACH(s, sessions) {
                _cleanup_free_ char *type = NULL, *tty = NULL, *seat = NULL, *user = NULL, *service = NULL, *class = NULL;

                if (sd_session_get_uid(*s, &uid) < 0 || uid == getuid())
                        continue;

                if (sd_session_get_class(*s, &class) < 0 || !streq(class, "user"))
                        continue;

                if (sd_session_get_type(*s, &type) < 0 || !STR_IN_SET(type, "x11", "wayland", "tty", "mir"))
                        continue;

                sd_session_get_tty(*s, &tty);
                sd_session_get_seat(*s, &seat);
                sd_session_get_service(*s, &service);
                user = uid_to_name(uid);

                log_warning("User %s is logged in on %s.", strna(user), isempty(tty) ? (isempty(seat) ? strna(service) : seat) : tty);
                c++;
        }
```

https://github.com/systemd/systemd/blob/f64222b748420b97d11638bf9c5c7d64ade6ca35/src/systemctl/systemctl-logind.c#L208

Again, for the sake of agreeing in terminology, this is not done through the inhibit protocol and it's not ssh activating an inhibit lock, this is a hardcoded check in systemd checking an active logind session with specific attributes.

**Comment by wolfgangkarall** [comment]
@hdante Thanks for pointing out those details, and trying to clear up the terminology, the 'ignore inhibitors' in the systemctl response seems like a red herring...

I still think the current title of the issue is misleading. Might be the title fits the issue that has been marked as duplicate (#30926), but this one wants the exact opposite, a way to ignore the blocking of suspend (by whatever the mechanism is, logind/inhibitor/whatever), not some 'specially marked session to act as inhibitor'.

Because even if @poettering is doubting that "ssh logins are already inhibiting suspend" and coming to the conclusion that "such a logic simply doesn't exist", from a user's perspective that is the case (in the non-root, different-user scenario described in my last comment, maybe in #30926 it's the local and remotely logged in user is the same).

**Comment by dtardon** [comment]
> Because even if @poettering is doubting that "ssh logins are already inhibiting suspend" and coming to the conclusion that "such a logic simply doesn't exist",

Well, it doesn't. This has nothing to do with ssh. It's a basic feature of a multi-user system: a user should not be able to power off the system if _other_ users are logged in (unless he's an admin). And adding a config. knob to logind to circumvent this logic is superfluous, as there already _is_ a way to do it: run the power off command with elevated privileges.

**Comment by tloimer** [comment]
My observation is that another logged-in user counts as "inhibitor", also for root:
~~~
root:/# systemctl --version
systemd 252 (252.17-1~deb12u1)
root:/# systemctl --dry-run --check-inhibitors=yes poweroff && echo success
User no-admin is logged in on sshd.
Please retry operation after closing inhibitors and logging out other users.
Alternatively, ignore inhibitors and users with 'systemctl poweroff -i'.
root:/# systemctl --dry-run poweroff && echo success
success
~~~
Root is able to power off the system, but only if it ignores inhibitors. A logind session should not be counted as inhibitor, as the above version of systemd seems to do.

**Comment by wolfgangkarall** [comment]
> > Because even if @poettering is doubting that "ssh logins are already inhibiting suspend" and coming to the conclusion that "such a logic simply doesn't exist",
> 
> Well, it doesn't. This has nothing to do with ssh. It's a basic feature of a multi-user system: a user should not be able to power off the system if _other_ users are logged in (unless he's an admin). And adding a config. knob to logind to circumvent this logic is superfluous, as there already _is_ a way to do it: run the power off command with elevated privileges.

In fact, independent of any other users being logged in, in a multi-user system a non-admin user shouldn't be allowed to power off or suspend the system at all IMO. And that was the default for a long time in Linux, nowadays it seems, 'console users' (i.e. ones that are using the system via some physical, local means) are allowed to do so.

But what is it exactly that you're suggesting? The this call to suspend the notebook, which works for the user without elevated privileges when there's no other user logged in via ssh, should be done _with_ elevated privileges all the time?

**Comment by hdante** [comment]
> > Because even if @poettering is doubting that "ssh logins are already inhibiting suspend" and coming to the conclusion that "such a logic simply doesn't exist",
> 
> Well, it doesn't. This has nothing to do with ssh. It's a basic feature of a multi-user system: a user should not be able to power off the system if _other_ users are logged in (unless he's an admin). And adding a config. knob to logind to circumvent this logic is superfluous, as there already _is_ a way to do it: run the power off command with elevated privileges.

Hello, the manual shutdown command was only used to isolate the problem. In the original description it was implicit suspend due to closing the lid.

**Comment by wolfgangkarall** [comment]
@poettering It's great to see that nobody ever cared about the original problem described after the issue was hijacked, so whatever you completed, it's not this issue.

Of course, even as the person that opened the issue and who is quite disagreeing about #42275 closing this issue at all,  I'm not allowed to reopen it.

**Comment by hdante** [comment]
Hello, the new patch recently merged solves my problem. Since my problem is exactly the opposite of that of the original post, I confirm that this issue has not been resolved. I'll quote again the original issue and it will become clear:

> **Is your feature request related to a problem? Please describe.**
> When closing my notebook's lid while e.g. my SSH-based backup is running (i.e. having an active SSH session) the notebook does not go to sleep.

As already discussed, the issue is caused by hard-coded disabling of suspend by logind (assuming no custom scripts are involved). Since the patch doesn't touch logind code, it's impossible for the patch to have solved the problem. The problem can only be solved by modifying the hard-coded logic, for example, by making it configurable, smarter, or deleting it.

This issue should be reopened, or a new issue created to report the same thing.
